### PR TITLE
Debounce and cache tooltips

### PR DESCRIPTION
**Pull request by unorsk**
I noticed that sometimes tooltips on Unison share are too slow :)
Wondering if something like that could be an Ok solution
I added a delay before even trying to load a tooltip – so it won't make an http request when I accidentally hover on a term. And then I also added some primitive caching so it won't load same term twice.

Before:

https://user-images.githubusercontent.com/25188/214340703-519e1431-ec80-4156-af51-a7302ffc4329.mov

Now:

https://user-images.githubusercontent.com/25188/214340818-bd5b857a-6c2c-4467-839b-ee18464182fd.mov

### CODE REVIEW

File: src/Code/DefinitionSummaryTooltip.elm [hunk 1]
```diff
@@ -29,48 +30,73 @@ type DefinitionSummary
...
+    | CachedDefinition Reference (WebData DefinitionSummary)
     | FetchDefinitionFinished Reference (HttpResult DefinitionSummary)
 
-
 update : Config -> Msg -> Model -> ( Model, Cmd Msg )
 update config msg model =
+    let debounceDelay = 500
+    in
     case msg of
+        ShowTooltipDelayed ref ->
```
Comment (hojberg) [inline src/Code/DefinitionSummaryTooltip.elm:64]: What do you think about having this be `ShowTooltip` and the other one be `FetchDefinition`

File: src/Code/DefinitionSummaryTooltip.elm [hunk 2]
```diff
@@ -29,48 +30,73 @@ type DefinitionSummary
...
+    = ShowTooltipDelayed Reference
+    | ShowTooltip Reference
     | HideTooltip Reference
+    | CachedDefinition Reference (WebData DefinitionSummary)
     | FetchDefinitionFinished Reference (HttpResult DefinitionSummary)
 
-
 update : Config -> Msg -> Model -> ( Model, Cmd Msg )
 update config msg model =
+    let debounceDelay = 500
```
Comment (hojberg) [inline src/Code/DefinitionSummaryTooltip.elm:61]: How does it feel with 300?

**Review by hojberg — COMMENTED** [review]
🎉 Really cool! Thank you so much for this :)

I've added a few suggestions.
Also (and I apologize there's no CI for this) could you run `elm-format` on it? Thank you.

File: src/Code/DefinitionSummaryTooltip.elm (hunk 1 above)
Comment (unorsk) [reply src/Code/DefinitionSummaryTooltip.elm]: Yeah, that makes sense!

File: src/Code/DefinitionSummaryTooltip.elm (hunk 2 above)
Comment (unorsk) [reply src/Code/DefinitionSummaryTooltip.elm]: I actually first tried 300 as it is used some other place in the code. But than bumped it up to 2000 when testing and slowly got down to 500 for some reason forgetting about the initial 300 🙈
But yeah, 300 feels snappier :)

File: src/Code/DefinitionSummaryTooltip.elm [hunk 3]
```diff
@@ -29,48 +30,73 @@ type DefinitionSummary
...
 
 -- UPDATE
 
 
 type Msg
-    = ShowTooltip Reference
+    = ShowTooltipDelayed Reference
+    | ShowTooltip Reference
     | HideTooltip Reference
+    | CachedDefinition Reference (WebData DefinitionSummary)
```
[the remark this answers is not in the document]
Comment (unorsk) [reply src/Code/DefinitionSummaryTooltip.elm]: Right, it's some leftovers I missed. Fixed now.

File: src/Code/DefinitionSummaryTooltip.elm [hunk 4]
```diff
@@ -29,48 +30,73 @@ type DefinitionSummary
     | DataConstructorHover DataConstructorSummary
     | AbilityConstructorHover AbilityConstructorSummary
 
-
 type alias Model =
-    Maybe ( Reference, WebData DefinitionSummary )
+    {
+        data: Maybe ( Reference, WebData DefinitionSummary )
+        , alreadyFetched: Dict String (Reference, WebData DefinitionSummary)
+    }
```
[the remark this answers is not in the document]
Comment (unorsk) [reply src/Code/DefinitionSummaryTooltip.elm:38]: I like the idea of making it more efficient!
But we needed WebData in `(activeTooltip : Maybe ( Reference, WebData DefinitionSummary ))` so we could track progress of the current http-request. What we probably don't need is storing WebData for all the summaries.

It should probably look more like this:
```
type alias Model =
  { activeRef : ( Reference, WebData DefinitionSummary )
  , summaries : Dict String DefinitionSummary
  }
```
What do you think, does this make sense?

And then we're using `Reference` in the view and I though that just swapping `activeTooltip` with an entry from the cache would be so much easier, this is how I ended up with storing everything :)

**Review by hojberg — APPROVED** [review]
:tada:

**Pull request merged**